R/auto.R

```diff
@@ -1,75 +1,144 @@
+  config <- lapply(cols, function(x) {
     if (isTRUE(is.na(x))) x else parse_any_color(x)
   })
   if (!is.null(font)) {
-    defaults$font <- as_font_spec(font)
+    config$font <- as_font_spec(font)
   }
-  oldPrefs <- .globals$auto_defaults
-  if (length(defaults)) {
-    .globals$auto_defaults <- defaults
+  if (!is.character(priority) || length(priority) < 1) {
+    stop("`priority` must be a character vector of length 1 or more", call. = FALSE)
+  }
+  missing_priority <- setdiff(priority, priorities())
+  if (length(missing_priority)) {
+    stop(
+      "Didn't recognize the following priority values: '",
+      paste(missing_priority, collapse = "', '"), "'", call. = FALSE
+    )
+  }
+  config$priority <- priority
+  structure(config, class = "thematic_auto_config")
+}
+
+#' @rdname auto-config
+#' @param config a `auto_config()` object.
+#' @export
+auto_config_set <- function(config) {
+  if (!inherits(config, "thematic_auto_config")) {
+    stop("`config` must be a `auto_config()` object", call. = FALSE)
   }
-  invisible(oldPrefs)
+  old_config <- auto_config_get()
+  .globals$auto_config <- config
+  invisible(old_config)
 }
 
+#' @rdname auto-config
+#' @export
+auto_config_get <- function() {
+  # This should contain priority even if a config hasn't been set
+  config <- utils::modifyList(auto_config(), .globals$auto_config %||% list())
+  structure(config, class = "thematic_auto_config")
+}
 
-resolve_auto_theme <- function() {
-  theme <- .globals$theme
-  outputInfo <- shiny_output_info()
-  autoDefaults <- auto_defaults()
-  bsThemeColors <- bs_theme_colors()
-  rsThemeColors <- rs_theme_colors()
+priorities <- function() {
+  eval(formals(auto_config)$priority)
+}
+
+#' Resolve auto values
+#'
+#' Resolves `'auto'` values based on the current execution environment
+#' and configuration (i.e., [auto_config_get()]).
+#'
+#' @return The `theme` object with resolved `'auto'` values.
+#'
+#' @param theme a `thematic_theme()` object.
+#' @export
+#' @seealso [auto_config_set()]
+#' @examples
+#'
+#' old_config <- auto_config_set(auto_config(bg = "black", fg = "white"))
+#'
+#' # Resolving auto values in local theme objects
+#' theme <- thematic_theme()
+#' theme[c("bg", "fg")]
+#' theme <- auto_resolve_theme(theme)
+#' theme[c("bg", "fg")]
+#'
+#' # By default, auto values are resolved when accessing
+#' # global theme options
+#' thematic_on()
+#' thematic_get_option("bg", resolve = FALSE)
+#' thematic_get_option("bg")
+#' thematic_off()
+#'
+#' auto_config_set(old_config)
+#'
+auto_resolve_theme <- function(theme) {
+  if (!length(theme)) {
+    return(theme)
+  }
+  if (!is_thematic_theme(theme)) {
+    stop("`theme` must be a `thematic_theme()` object", call. = FALSE)
+  }
+
+  auto_color_info <- list(
+    shiny = shiny_output_info(),
+    config = auto_config_get(),
+    bootstraplib = bs_theme_colors(),
+    rstudio = rs_theme_colors()
+  )
+  auto_color_info <- auto_color_info[auto_config_get()$priority]
```

Comment:
Could check for sensible priority values here

Reply:
Also consider using `lapply(priority, function(x) switch(x, ..., stop("unknown")))`